tags/views.py

```diff
@@ -0,0 +1,115 @@
+from django.http import HttpResponseBadRequest, HttpResponse, JsonResponse
+from django_filters.rest_framework import DjangoFilterBackend
+from rest_framework import mixins
+from rest_framework import viewsets
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.views import APIView
+from rest_framework.viewsets import GenericViewSet
+
+from tags.models import Namespace, Tag
+from tags.permissions import (
+    NamespacePermission,
+    user_can_link_tag_to,
+    ManageTagPermission,
+)
+from tags.serializers import NamespaceSerializer, TagSerializer
+
+"""
+
+To create a namespace : 
+- either be an admin and create a global scope
+- either specify a scope and a scoped_to and have permissions on the scoped_to object
+
+"""
+
+
+class NamespaceViewSet(viewsets.ModelViewSet):
+    queryset = Namespace.objects.all()
+    serializer_class = NamespaceSerializer
+    filter_backends = (DjangoFilterBackend,)
+    filterset_fields = ("scope", "scoped_to")
+    permission_classes = (NamespacePermission,)
+
+    def create(self, request, *args, **kwargs):
+        # Required since we use "unique_together"
+        if "scoped_to" not in request.data:
+            request.data["scoped_to"] = ""
+
+        return super(NamespaceViewSet, self).create(request, *args, **kwargs)
+
+    def update(self, request, *args, **kwargs):
+        if "scope" in request.data or "scoped_to" in request.data:
+            return HttpResponseBadRequest("Cannot change scope of a namespace")
+
+        return super(NamespaceViewSet, self).update(request, *args, **kwargs)
+
+
+class TagViewSet(
+    mixins.CreateModelMixin,
+    mixins.DestroyModelMixin,
+    mixins.ListModelMixin,
+    GenericViewSet,
+):
+    serializer_class = TagSerializer
+    permission_classes = (ManageTagPermission,)
+    queryset = Tag.objects.all()
+
+    def get_queryset(self):
+        queryset = self.queryset
+
+        scope = self.request.query_params.get("scope", None)
+        scope_id = self.request.query_params.get("scope_id", None)
+
+        if scope is not None and scope_id is not None:
+            queryset = queryset.filter(
+                namespace__scope=scope, namespace__scoped_to=scope_id
+            )
+
+        return queryset
+
+    def create(self, request, *args, **kwargs):
+        if "namespace" in request.data and "value" in request.data:
+            try:
+                tag = Tag.objects.get(
+                    namespace=request.data.get("namespace"),
+                    value=request.data.get("value"),
+                )
+                return JsonResponse(TagSerializer().to_representation(tag), status=201)
```

Comment:
I think it's better to use the DRF object `rest_framework.response.Response`, at least we are sure it will integrate well with the DRF framework (in particular the browsable API)
I saw you imported `django.HttpResponse`, same for this one